prow/cmd/moonraker/main.go

```diff
@@ -0,0 +1,133 @@
+	"k8s.io/test-infra/prow/interrupts"
+	"k8s.io/test-infra/prow/logrusutil"
+	"k8s.io/test-infra/prow/metrics"
+	"k8s.io/test-infra/prow/moonraker"
+	"k8s.io/test-infra/prow/pjutil"
+)
+
+type options struct {
+	client         prowflagutil.KubernetesOptions
+	github         prowflagutil.GitHubOptions
+	port           int
+	cookiefilePath string
+
+	config configflagutil.ConfigOptions
+
+	dryRun                 bool
+	gracePeriod            time.Duration
+	instrumentationOptions prowflagutil.InstrumentationOptions
+}
+
+func gatherOptions(fs *flag.FlagSet, args ...string) options {
+	var o options
+	fs.IntVar(&o.port, "port", 8080, "HTTP port.")
+	fs.DurationVar(&o.gracePeriod, "grace-period", 180*time.Second, "On shutdown, try to handle remaining events for the specified duration. ")
+	fs.StringVar(&o.cookiefilePath, "cookiefile", "", "Path to git http.cookiefile, leave empty for github or anonymous")
+	for _, group := range []flagutil.OptionGroup{&o.client, &o.github, &o.instrumentationOptions, &o.config} {
+		group.AddFlags(fs)
+	}
+
+	fs.Parse(args)
+
+	return o
+}
+
+func (o *options) validate() error {
+	var errs []error
+	for _, group := range []flagutil.OptionGroup{&o.client, &o.github, &o.instrumentationOptions, &o.config} {
+		if err := group.Validate(o.dryRun); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	return utilerrors.NewAggregate(errs)
+}
+
+func main() {
+	logrusutil.ComponentInit()
+
+	o := gatherOptions(flag.NewFlagSet(os.Args[0], flag.ExitOnError), os.Args[1:]...)
+	if err := o.validate(); err != nil {
+		logrus.WithError(err).Fatal("Invalid options")
+	}
+
+	configAgent, err := o.config.ConfigAgent()
+	if err != nil {
+		logrus.WithError(err).Fatal("Error starting config agent.")
+	}
+
+	// If we are provided credentials for Git hosts, use them. These credentials
+	// hold per-host information in them so it's safe to set them globally.
+	if o.cookiefilePath != "" {
+		cmd := exec.Command("git", "config", "--global", "http.cookiefile", o.cookiefilePath)
+		if err := cmd.Run(); err != nil {
+			logrus.WithError(err).Fatal("unable to set cookiefile")
+		}
+	}
+
+	gitClient, err := o.github.GitClientFactory(o.cookiefilePath, &o.config.InRepoConfigCacheDirBase, o.dryRun, false)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error getting Git client.")
+	}
+
+	cacheGetter, err := config.NewInRepoConfigCache(o.config.InRepoConfigCacheSize, configAgent, gitClient)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error creating InRepoConfigCacheGetter.")
+	}
+	metrics.ExposeMetrics("moonraker", configAgent.Config().PushGateway, o.instrumentationOptions.MetricsPort)
+
+	// Start serving liveness endpoint /healthz.
+	health := pjutil.NewHealthOnPort(o.instrumentationOptions.HealthPort)
+
+	mr := moonraker.Moonraker{
+		ConfigAgent:       configAgent,
+		InRepoConfigCache: cacheGetter,
+	}
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/get-inrepoconfig", mr.ServeGetInrepoconfig)
```

Comment:
nit: GET is already an HTTP verb so `/inrepoconfig` is probably more idiomatic.

Reply:
We actually do a POST to this endpoint with the payload (`http.MethodPost` in the Moonraker client). I included the `get-` in here because I wanted to make our position of "we are not following REST at all" very loud and clear (RESTfulness means we use nouns for the paths because the verbs live in the HTTP verbs GET, POST, etc.).

Also see https://stackoverflow.com/a/983458/437583, which states that GET + body is not a good idea for multiple reasons.

Still, I am OK with dropping the leading `get-` if you'd rather prefer that style. Thoughts?

Reply:
I prefer dropping it in favor of a noun, but it doesn't matter too much. Just naming.

Reply:
I don't have a strong opinion here because we're not following any conventions in either case. Will update to the noun-only name.